cmd/goat/account_migrate.go

```diff
@@ -0,0 +1,263 @@
+
+	// login client on the new host
+	sess, err := comatproto.ServerCreateSession(ctx, &newClient, &comatproto.ServerCreateSession_Input{
+		Identifier: did,
+		Password:   newPassword,
+	})
+	if err != nil {
+		return fmt.Errorf("failed login to newly created account on new host: %w", err)
+	}
+	newClient.Auth = &xrpc.AuthInfo{
+		Did:        did,
+		AccessJwt:  sess.AccessJwt,
+		RefreshJwt: sess.RefreshJwt,
+	}
+
+	// 2. Migrate Data
+
+	slog.Info("migrating repo")
+	repoBytes, err := comatproto.SyncGetRepo(ctx, oldClient, did, "")
+	if err != nil {
+		return fmt.Errorf("failed exporting repo: %w", err)
+	}
+	err = comatproto.RepoImportRepo(ctx, &newClient, bytes.NewReader(repoBytes))
+	if err != nil {
+		return fmt.Errorf("failed importing repo: %w", err)
+	}
+
+	slog.Info("migrating preferences")
+	// TODO: service proxy header for app.bsky?
+	prefResp, err := appbsky.ActorGetPreferences(ctx, oldClient)
+	if err != nil {
+		return fmt.Errorf("failed fetching old preferences: %w", err)
+	}
+	err = appbsky.ActorPutPreferences(ctx, &newClient, &appbsky.ActorPutPreferences_Input{
+		Preferences: prefResp.Preferences,
+	})
+	if err != nil {
+		return fmt.Errorf("failed importing preferences: %w", err)
+	}
+
+	slog.Info("migrating blobs")
+	blobCursor := ""
+	for {
+		listResp, err := comatproto.SyncListBlobs(ctx, oldClient, blobCursor, did, 100, "")
+		if err != nil {
+			return fmt.Errorf("failed listing blobs: %w", err)
+		}
+		for _, blobCID := range listResp.Cids {
+			blobBytes, err := comatproto.SyncGetBlob(ctx, oldClient, blobCID, did)
+			if err != nil {
+				slog.Warn("failed downloading blob", "cid", blobCID, "err", err)
+				continue
+			}
+			_, err = comatproto.RepoUploadBlob(ctx, &newClient, bytes.NewReader(blobBytes))
+			if err != nil {
+				slog.Warn("failed uploading blob", "cid", blobCID, "err", err, "size", len(blobBytes))
+			}
+			slog.Info("transfered blob", "cid", blobCID, "size", len(blobBytes))
+		}
+		if listResp.Cursor == nil || *listResp.Cursor == "" {
+			break
+		}
+		blobCursor = *listResp.Cursor
+	}
+
+	// display migration status
+	// TODO: should this loop? with a delay?
+	statusResp, err := comatproto.ServerCheckAccountStatus(ctx, &newClient)
+	if err != nil {
+		return fmt.Errorf("failed checking account status: %w", err)
+	}
+	slog.Info("account migration status", "status", statusResp)
+
+	// 3. Migrate Identity
+	// TODO: support did:web etc
+	slog.Info("updating identity to new host")
+
+	credsResp, err := IdentityGetRecommendedDidCredentials(ctx, &newClient)
+	if err != nil {
+		return fmt.Errorf("failed fetching new credentials: %w", err)
+	}
+	credsBytes, err := json.Marshal(credsResp)
+	if err != nil {
+		return nil
+	}
+
+	var unsignedOp IdentitySignPlcOperation_Input
+	if err = json.Unmarshal(credsBytes, &unsignedOp); err != nil {
+		return fmt.Errorf("failed parsing PLC op: %w", err)
+	}
+	unsignedOp.Token = &plcToken
+
+	// TODO: add some safety checks here around rotation key set intersection?
```

Comment:
The new PDS does some checks on your behalf as well. But an extra check here wouldn't hurt. The TS script I wrote doesn't do any extra checks but before we start encouraging other folks to use this script en mass we should probably make it a _bit_ more resilient